File: app/src/main/java/net/squanchy/speaker/widget/SpeakerHeaderView.java
package net.squanchy.speaker.widget;

import android.content.Context;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import net.squanchy.R;
import net.squanchy.imageloader.ImageLoader;
import net.squanchy.imageloader.ImageLoaderInjector;
import net.squanchy.speaker.domain.view.Speaker;
import net.squanchy.support.lang.Optional;

public class SpeakerHeaderView extends LinearLayout {

    @Nullable
    private ImageLoader imageLoader;

    private ImageView photoView;
    private TextView nameView;
    private TextView companyView;

    public SpeakerHeaderView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public SpeakerHeaderView(Context context, AttributeSet attrs, int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }

    public SpeakerHeaderView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);

        if (!isInEditMode()) {
            imageLoader = ImageLoaderInjector.obtain(context).imageLoader();
        }

        super.setOrientation(HORIZONTAL);
    }

    @Override
    public void setOrientation(int orientation) {
        throw new UnsupportedOperationException("Changing orientation is not supported for SpeakerHeaderView");
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        photoView = (ImageView) findViewById(R.id.speaker_photo);
        nameView = (TextView) findViewById(R.id.speaker_name);
        companyView = (TextView) findViewById(R.id.speaker_company);
    }

    public void updateWith(Speaker speaker) {
        updatePhoto(speaker.photoUrl());

        nameView.setText(speaker.name());

        Optional<String> companyName = speaker.companyName();
        if (companyName.isPresent()) {
            // TODO support navigating to company website
            companyView.setText(companyName.get());
            companyView.setVisibility(VISIBLE);
        } else {
            companyView.setVisibility(GONE);
        }
    }

    private void updatePhoto(Optional<String> photoUrl) {
        if (imageLoader == null) {
            throw new IllegalStateException("Unable to access the ImageLoader, it hasn't been initialized yet");
        }

        if (photoUrl.isPresent()) {
            photoView.setVisibility(VISIBLE);
            imageLoader.load(photoUrl.get()).into(photoView);
        } else {
            photoView.setVisibility(GONE);
        }
    }
}

File: app/src/main/java/net/squanchy/speaker/widget/SpeakerHeaderView.kt
package net.squanchy.speaker.widget

import android.content.Context
import android.util.AttributeSet
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.view.isVisible
import arrow.core.Option
import kotlinx.android.synthetic.main.activity_speaker_details.view.*
import net.squanchy.R
import net.squanchy.imageloader.ImageLoader
import net.squanchy.imageloader.imageLoaderComponent
import net.squanchy.speaker.domain.view.Speaker
import net.squanchy.support.lang.getOrThrow
import net.squanchy.support.unwrapToActivityContext

class SpeakerHeaderView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet,
    defStyleAttr: Int = 0
) : ConstraintLayout(context, attrs, defStyleAttr) {

    private var imageLoader: ImageLoader? = null

    init {
        if (!isInEditMode) {
            val activity = context.unwrapToActivityContext()
            imageLoader = imageLoaderComponent(activity).imageLoader()
        }
    }

    fun updateWith(speaker: Speaker) {
        updatePhoto(speaker.photoUrl, imageLoader)

        speakerNames.text = speaker.name

        val companyName = speaker.companyName
        if (companyName.isDefined()) {
            // TODO support navigating to company website
            speakerCompany.text = companyName.getOrThrow()
            speakerCompany.isVisible = true
        } else {
            speakerCompany.isVisible = false
        }
    }

    private fun updatePhoto(photoUrl: Option<String>, imageLoader: ImageLoader?) {
        if (imageLoader == null) {
            throw IllegalStateException("Unable to access the ImageLoader, it hasn't been initialized yet")
        }

        if (photoUrl.isDefined()) {
            speakerPhoto.isVisible = true
            imageLoader.load(photoUrl.getOrThrow())
                .error(R.drawable.ic_no_avatar)
                .into(speakerPhoto)
        } else {
            speakerPhoto.setImageResource(R.drawable.ic_no_avatar)
        }
    }
}
